clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -402,27 +403,41 @@ public void refreshCommittedOffsetsIfNeeded() {
         }
     }
 
-    @Override
-    public void close() {
+    public void close(long timeoutMs, long requestTimeoutMs) {
         // we do not need to re-enable wakeups since we are closing already
         client.disableWakeups();
+
+        long now = time.milliseconds();
+        timeoutMs = Math.min(timeoutMs, requestTimeoutMs);
+        long endTimeMs = now + timeoutMs;
         try {
-            maybeAutoCommitOffsetsSync();
-
-            Node coordinator;
-            long endTimeMs = time.milliseconds() + CLOSE_TIMEOUT_MS;
-            while ((coordinator = coordinator()) != null && client.pendingRequestCount(coordinator) > 0) {
-                long remainingTimeMs = endTimeMs - time.milliseconds();
-                if (remainingTimeMs > 0)
-                    client.poll(remainingTimeMs);
-                else {
-                    log.warn("Close timed out with {} pending requests to coordinator, terminating client connections for group {}.", client.pendingRequestCount(coordinator), groupId);
-                    break;
-                }
+            maybeAutoCommitOffsetsSync(timeoutMs);
+            now = time.milliseconds();
+            if (pendingAsyncCommits.get() > 0) {
```

Comment:
`pendingAsyncCommits` now indicates that the async commit has not yet been queued (i.e. it is still waiting to find the coordinator). That brings async commits to the same state as auto-commits. Commits already queued are handled by the subsequent loop (i.e. process request if coordinator != null).

`maybeAutoCommitOffsetsSync` already does `ensureCoordinatorReady` if there are offsets to commit. And since `LeaveGroup` was already on a best-effort basis (only send if coordinator was known), I thought it was better to keep it as-is to avoid delays in shutdown.